src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -2989,11 +2989,27 @@ internal static bool CheckValEscape(SyntaxNode node, BoundExpression expr, uint
             {
                 case BoundKind.DefaultLiteral:
                 case BoundKind.DefaultExpression:
-                case BoundKind.Parameter:
                 case BoundKind.ThisReference:
                     // always returnable
                     return true;
 
+                case BoundKind.Parameter:
+                    var parameter = ((BoundParameter)expr).ParameterSymbol;
+                    // params Span<T> value cannot be returned since that would
+                    // prevent sharing repeated allocations at the call-site.
+                    if (parameter.IsParams &&
+                        parameter.Ordinal == parameter.ContainingSymbol.GetParameterCount() - 1 &&
+                        !parameter.Type.IsSZArray())
```

Comment:
>!parameter.Type.IsSZArray() [](http://example.com/codeflow?start=24&length=27)

If we are complaining about a span type, then we probably should check for that rather than checking that the type is not an array.

Reply:
I think in error situations the type could be neither and the error will be unexpected and confusing.

Reply:
Added PROTOTYPE comment.